### PR TITLE
SetReversePInvokeFrameSlot into gcInfo for reversePInvoke

### PR DESCRIPTION
**Pull request by sandreenko**
Fix dotnet/corert#2115.

### CODE REVIEW

**Comment by sandreenko** [comment]
@dotnet-bot test Linux ARM Emulator Cross Debug Build  please

**Comment by sandreenko** [comment]
@pgavlin  is it fine to do it unconditionally or should  it be surrounded with if CORERT_ABI?
I don't see how it can non-CoreRT compilers.

File: src/jit/codegenxarch.cpp [hunk 1]
```diff
@@ -8380,6 +8380,11 @@ void CodeGen::genCreateAndStoreGCInfoX64(unsigned codeSize, unsigned prologSize
         // frame
         gcInfoEncoder->SetSizeOfEditAndContinuePreservedArea(preservedAreaSize);
     }
+    if (compiler->opts.IsReversePInvoke())
+    {
+        assert(compiler->lvaReversePInvokeFrameVar != BAD_VAR_NUM);
+        gcInfoEncoder->SetReversePInvokeFrameSlot(compiler->lvaReversePInvokeFrameVar);
```
Comment (jkotas) [inline src/jit/codegenxarch.cpp]: This needs to be spOffset, not the variable number.

**Comment by pgavlin** [comment]
@sandreenko: I do think that this should be conditional on a CoreRT ABI check, yes.

File: src/jit/codegenxarch.cpp (hunk 1 above)
Comment (pgavlin) [inline src/jit/codegenxarch.cpp]: Something like `compiler->lvaTable[compiler->lvaReversePInvokeFrameVar].lvStkOffs` should be sufficient to retrieve the offset of the variable on the frame. IIRC, however, that will be the offset from the frame pointer (whether it is in rbp or rsp). @BruceForstall can you confirm?

**Comment by jkotas** [comment]
It is conditional for CoreRT indirectly via IsReversePInvoke. IsReversePInvoke is only ever set for CoreRT today. If we ever use it for CoreCLR as well, we will want to have the offset recorded in GCInfo as well. I think that the current check for IsReversePInvoke is sufficient and correct.

File: src/jit/codegenxarch.cpp (hunk 1 above)
Comment (jkotas) [inline src/jit/codegenxarch.cpp]: > however, that will be the offset from the frame pointer (whether it is in rbp or rsp

That's ok - as long as it matches how we interpret the offset at runtime.

Comment (sandreenko) [inline src/jit/codegenxarch.cpp]: FIxed

File: src/jit/codegenxarch.cpp [hunk 2]
```diff
@@ -8381,6 +8381,14 @@ void CodeGen::genCreateAndStoreGCInfoX64(unsigned codeSize, unsigned prologSize
         gcInfoEncoder->SetSizeOfEditAndContinuePreservedArea(preservedAreaSize);
     }
 
+    if (compiler->opts.IsReversePInvoke())
+    {
+        unsigned reversePInvokeFrameVarNumber = compiler->lvaReversePInvokeFrameVar;
+        assert(reversePInvokeFrameVarNumber != BAD_VAR_NUM && reversePInvokeFrameVarNumber < compiler->lvaRefCount);
+        LclVarDsc reversePInvokeFrameVar = compiler->lvaTable[reversePInvokeFrameVarNumber];
```
Comment (pgavlin) [inline src/jit/codegenxarch.cpp]: Generally we use `LclVarDsc* var = &compiler->lvaTable[offset]` to avoid copying the entire `LclVarDsc`.

Comment (sandreenko) [inline src/jit/codegenxarch.cpp]: Fixed

**Pull request merged**